Improve ClientError  handling

ValentinBuira opened the pull request:
This PR improve the ClientError exception to make it easier to parse and return feedback to user.  Instead of relying on a text based error we define variable member of the ClientError Exception than will hold the data response from the server. 

Used only for the qgis-plugin for now

wonder-sk commented on mergin/client.py:
```diff
@@ -205,19 +205,26 @@ def _do_request(self, request):
+            server_response = None
+            if e.headers.get("Content-Type", "") == "application/json":
+                server_response = json.load(e)
+                err_detail = server_response.get("detail")
```
Is this a valid thing to do (i.e. shall we expect "detail" here)? Do we have some example where server returns `application/json` content type? I thought that if there's `detail`, then it will be marked as `application/problem+json` ?

wonder-sk commented on mergin/client.py:
```diff
@@ -205,19 +205,26 @@ def _do_request(self, request):
+            if e.headers.get("Content-Type", "") == "application/json":
+                server_response = json.load(e)
+                err_detail = server_response.get("detail")
+                server_code = server_response.get("code")
```
How does it work - the `code` is only transferred in this case, but not in `application/problem+json` case?

(probably I am just not familiar enough with various server error responses :slightly_smiling_face: )

tomasMizera replied:
We checked the server side and it seems we are returning simple `application/json` and not the one with `problem+json`. Here is a ticket https://github.com/MerginMaps/server-private/issues/2459, I'd prefer if we did not check the `Content-Type` for now. The mobile app does not do it either.

wonder-sk reviewed (commented):
It would be good to add also some auto tests that check for some common errors, making sure ClientError is populated as expected...

ValentinBuira commented:
I updated the PR with the requested changes, and it's ready for another review,  to sum up:

* Created two new tests `test_error_push_already_named_project` and `test_error_projects_limit_hit`
* Added a utility function `bytes_to_human_size` (used in the plugin)
* Always parse json response from the server the same way whatever content-type is the response

Thanks for your patience

MarcelGeo commented on mergin/test/test_client.py:
```diff
@@ -2629,3 +2630,37 @@ def test_editor_push(mc: MerginClient, mc2: MerginClient):
+    project = API_USER + "/" + test_project
+    project_dir = os.path.join(TMP_DIR, test_project)
+
+    try:
```
I think we can use this testing method to ensure that code raises exception:

```
# attempt to push project in the unfinished pull state should
    # fail with ClientError
    with pytest.raises(ClientError):
        mc.push_project(project_dir)
```

MarcelGeo reviewed (changes requested):
Just enhancements about tests
